test.cpp

```diff
@@ -380,6 +380,195 @@ TEST(AbspathTest, TrailingDotInSegment)
         Url::Url("http://foo.com//foo/whiz./bar").abspath().path());
 }
 
+TEST(RelativeTest, SchemeRelative)
+{
+    Url::Url base("http://foo.com/");
+    EXPECT_EQ("http://relative.com/",
+        Url::Url("//relative.com/").relative_to(base).str());
+}
+
+TEST(RelativeTest, IncludeUserinfo)
+{
+    Url::Url base("http://user@foo.com/");
+    EXPECT_EQ("http://user@foo.com/path",
+        Url::Url("/path").relative_to(base).str());
+}
+
+TEST(RelativeTest, AbsoluteRelative)
+{
+    Url::Url base("http://foo.com/");
+    EXPECT_EQ("https://relative.com/",
+        Url::Url("https://relative.com/").relative_to(base).str());
+}
+
+TEST(RelativeTest, RelativePath)
+{
+    Url::Url base("http://foo.com/a/b/c");
+    EXPECT_EQ("http://foo.com/a/b/some/path",
+        Url::Url("some/path").relative_to(base).str());
+}
+
+TEST(RelativeTest, RelativePathDirectory)
+{
+    Url::Url base("http://foo.com/a/b/c");
+    EXPECT_EQ("http://foo.com/a/b/some/path/",
+        Url::Url("some/path/").relative_to(base).str());
+}
+
+TEST(RelativeTest, AbsolutePath)
+{
+    Url::Url base("http://foo.com/a/b/c");
+    EXPECT_EQ("http://foo.com/some/path",
+        Url::Url("/some/path").relative_to(base).str());
+}
+
+TEST(RelativeTest, ParamsOnly)
+{
+    Url::Url base("http://foo.com/a/b/c;existing?query");
+    EXPECT_EQ("http://foo.com/a/b/;params",
+        Url::Url(";params").relative_to(base).str());
+}
+
+TEST(RelativeTest, PathAndParams)
+{
+    Url::Url base("http://foo.com/a/b/c");
+    EXPECT_EQ("http://foo.com/a/b/path;params",
+        Url::Url("path;params").relative_to(base).str());
+}
+
+TEST(RelativeTest, ParamsAndQuery)
+{
+    Url::Url base("http://foo.com/a/b/c");
+    EXPECT_EQ("http://foo.com/a/b/;params?query",
+        Url::Url(";params?query").relative_to(base).str());
+}
+
+TEST(RelativeTest, PathAndQuery)
+{
+    Url::Url base("http://foo.com/a/b/c");
+    EXPECT_EQ("http://foo.com/a/b/path?query",
+        Url::Url("path?query").relative_to(base).str());
+}
+
+TEST(RelativeTest, QueryOnly)
+{
+    Url::Url base("http://foo.com/a/b/c;params?existin");
+    EXPECT_EQ("http://foo.com/a/b/c;params?query",
+        Url::Url("?query").relative_to(base).str());
+}
+
+TEST(RelativeTest, PathAndFragment)
+{
+    Url::Url base("http://foo.com/a/b/c");
+    EXPECT_EQ("http://foo.com/a/b/path#fragment",
+        Url::Url("path#fragment").relative_to(base).str());
+}
+
+TEST(RelativeTest, FragmentOnly)
+{
+    Url::Url base("http://foo.com/a/b/c");
+    EXPECT_EQ("http://foo.com/a/b/c#fragment",
+        Url::Url("#fragment").relative_to(base).str());
+}
+
+TEST(RelativeTest, Dot)
```

Comment:
Can you add a test like this one with a base of `"http://foo.com/a/b/c/d/"` to confirm a trailing slash won't strip a directory with dot? Same thing for the `..` case.